File: baba-chatbot-python/src/baba_chatbot/integrations/llm/llm_client.py
"""
Cliente para interactuar con modelos de lenguaje de OpenAI.
Implementa generación de respuestas con RAG y function calling para búsqueda de vehículos.
"""
import json
import logging
from typing import List, Optional, Any, Dict
from openai import AsyncOpenAI
from ...application.abstractions.interfaces import ILlmClient, ICatalogRepository
from ...domain.value_objects import VehicleQuery
from .knowledge_repository import KnowledgeRepository


logger = logging.getLogger(__name__)


class LlmClient(ILlmClient):
    """
    Cliente para interactuar con modelos de lenguaje de OpenAI.
    """

    def __init__(
        self,
        api_key: str,
        model: str = "gpt-4o-mini",
        temperature: float = 0.7,
        knowledge_repository: Optional[KnowledgeRepository] = None,
        catalog_repository: Optional[ICatalogRepository] = None
    ):
        """
        Inicializa el cliente LLM.
        
        Args:
            api_key: API key de OpenAI
            model: Modelo a utilizar
            temperature: Temperatura para generación (0-1)
            knowledge_repository: Repositorio de conocimiento para RAG
            catalog_repository: Repositorio de catálogo para function calling
        """
        self._client = AsyncOpenAI(api_key=api_key)
        self._model = model
        self._temperature = temperature
        self._knowledge_repository = knowledge_repository
        self._catalog_repository = catalog_repository
        
        logger.info(f"LlmClient initialized with model: {model}")

    async def generate_response(
        self,
        system_prompt: str,
        user_message: str,
        conversation_history: Optional[List[str]] = None
    ) -> str:
        """
        Genera una respuesta usando el LLM con contexto RAG y herramientas de búsqueda de vehículos.
        Enriquece el prompt del sistema con información relevante de la base de conocimiento.
        
        Args:
            system_prompt: Prompt del sistema con instrucciones
            user_message: Mensaje del usuario
            conversation_history: Historial de conversación opcional
            
        Returns:
            Respuesta generada por el LLM
        """
        try:
            # Buscar contexto relevante en RAG si está disponible
            relevant_context = ""
            if self._knowledge_repository:
                relevant_context = self._knowledge_repository.search_relevant_context(
                    user_message,
                    max_chunks=3
                )

            # Enriquecer el prompt del sistema con contexto RAG
            enhanced_system_prompt = system_prompt
            if relevant_context:
                enhanced_system_prompt = (
                    f"{system_prompt}\n\n---\n\n"
                    "## INFORMACIÓN DE CONTEXTO (Base de Conocimiento)\n\n"
                    "Usa esta información para responder con datos precisos:\n\n"
                    f"{relevant_context}\n\n---\n\n"
                    "IMPORTANTE: Usa SOLO la información del contexto cuando esté disponible. "
                    "Si la información no está en el contexto, indica que necesitas verificar con un asesor."
                )
                logger.info(f"Enhanced prompt with {len(relevant_context)} characters of context")

            # Construir mensajes
            messages = [
                {"role": "system", "content": enhanced_system_prompt}
            ]

            # Agregar historial de conversación si existe
            if conversation_history:
                for i, msg in enumerate(conversation_history):
                    role = "user" if i % 2 == 0 else "assistant"
                    messages.append({"role": role, "content": msg})

            messages.append({"role": "user", "content": user_message})

            logger.debug(f"Sending chat completion request with {len(messages)} messages")

            # Preparar herramientas (function calling) si hay catálogo disponible
            tools = []
            if self._catalog_repository:
                tools = [
                    {
                        "type": "function",
                        "function": {
                            "name": "search_vehicles",
                            "description": "Busca vehículos en el inventario según criterios específicos como marca, modelo, precio, año, etc.",
                            "parameters": {
                                "type": "object",
                                "properties": {
                                    "brand": {
                                        "type": "string",
                                        "description": "Marca del vehículo (ej: Toyota, Honda, BMW)"
                                    },
                                    "model": {
                                        "type": "string",
                                        "description": "Modelo del vehículo (ej: Corolla, Civic, Serie 3)"
                                    },
                                    "min_price": {
                                        "type": "number",
                                        "description": "Precio mínimo en pesos"
                                    },
                                    "max_price": {
                                        "type": "number",
                                        "description": "Precio máximo en pesos"
                                    },
                                    "min_year": {
                                        "type": "integer",
                                        "description": "Año mínimo del vehículo"
                                    },
                                    "max_year": {
                                        "type": "integer",
                                        "description": "Año máximo del vehículo"
                                    },
                                    "max_mileage": {
                                        "type": "integer",
                                        "description": "Kilometraje máximo permitido"
                                    }
                                },
                                "required": []
                            }
                        }
                    },
                    {
                        "type": "function",
                        "function": {
                            "name": "get_vehicle_details",
                            "description": "Obtiene información detallada de un vehículo específico usando su ID o stock_id",
                            "parameters": {
                                "type": "object",
                                "properties": {
                                    "vehicle_id": {
                                        "type": "string",
                                        "description": "ID o stock_id del vehículo"
                                    }
                                },
                                "required": ["vehicle_id"]
                            }
                        }
                    }
                ]

            # Llamar a la API de OpenAI
            completion_args = {
                "model": self._model,
                "messages": messages,
                "temperature": self._temperature
            }
            
            if tools:
                completion_args["tools"] = tools

            completion = await self._client.chat.completions.create(**completion_args)

            # Verificar si hay llamadas a funciones
            if completion.choices[0].finish_reason == "tool_calls":
                response = await self._handle_tool_calls(messages, completion)
                return response

            response_text = completion.choices[0].message.content
            
            logger.info(f"Generated response: {len(response_text)} characters")
            
            return response_text
        except Exception as ex:
            logger.error(f"Error generating LLM response: {ex}")
            raise

    async def _handle_tool_calls(
        self,
        messages: List[Dict[str, Any]],
        completion: Any
    ) -> str:
        """
        Maneja las llamadas a herramientas (function calling) del LLM.
        Ejecuta las funciones solicitadas y envía los resultados de vuelta al modelo.
        
        Args:
            messages: Lista de mensajes de la conversación
            completion: Objeto de completion con las tool calls
            
        Returns:
            Respuesta final del LLM después de procesar las tool calls
        """
        if not self._catalog_repository:
            return "Lo siento, no puedo consultar el catálogo en este momento."

        # Agregar el mensaje del asistente con las tool calls
        assistant_message = completion.choices[0].message
        messages.append({
            "role": "assistant",
            "content": assistant_message.content,
            "tool_calls": [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {
                        "name": tc.function.name,
                        "arguments": tc.function.arguments
                    }
                }
                for tc in assistant_message.tool_calls
            ]
        })

        # Ejecutar cada tool call
        for tool_call in assistant_message.tool_calls:
            function_name = tool_call.function.name
            function_args = tool_call.function.arguments

            logger.info(f"Processing tool call: {function_name} with args: {function_args}")

            try:
                function_result = await self._execute_tool(function_name, function_args)
            except Exception as ex:
                logger.error(f"Error executing tool {function_name}: {ex}")
                function_result = json.dumps({"error": "Error ejecutando la búsqueda"})

            # Agregar resultado de la tool call
            messages.append({
                "role": "tool",
                "tool_call_id": tool_call.id,
                "content": function_result
            })

        # Obtener respuesta final del LLM
        final_completion = await self._client.chat.completions.create(
            model=self._model,
            messages=messages,
            temperature=self._temperature
        )

        return final_completion.choices[0].message.content

    async def _execute_tool(self, function_name: str, function_args: str) -> str:
        """
        Ejecuta una función/herramienta específica según su nombre.
        Soporta búsqueda de vehículos y obtención de detalles.
        
        Args:
            function_name: Nombre de la función a ejecutar
            function_args: Argumentos de la función en formato JSON
            
        Returns:
            Resultado de la función en formato JSON
        """
        if not self._catalog_repository:
            return json.dumps({"error": "Catálogo no disponible"})

        try:
            if function_name == "search_vehicles":
                params = json.loads(function_args)
                
                query = VehicleQuery(
                    brand=params.get("brand"),
                    model=params.get("model"),
                    min_price=params.get("min_price"),
                    max_price=params.get("max_price"),
                    min_year=params.get("min_year"),
                    max_year=params.get("max_year"),
                    max_mileage=params.get("max_mileage")
                )

                vehicles = await self._catalog_repository.search_vehicles(query)
                
                # Limitar a 10 resultados
                limited_vehicles = [
                    {
                        "id": v.stock_id,
                        "marca": v.brand,
                        "modelo": v.model,
                        "año": v.year,
                        "version": v.version,
                        "precio": v.price,
                        "kilometraje": v.mileage,
                        "dimensiones": {
                            "largo": v.length,
                            "ancho": v.width,
                            "alto": v.height
                        },
                        "caracteristicas": {
                            "bluetooth": v.has_bluetooth,
                            "carplay": v.has_carplay
                        }
                    }
                    for v in vehicles[:10]
                ]

                return json.dumps({
                    "total": len(vehicles),
                    "resultados_mostrados": len(limited_vehicles),
                    "vehiculos": limited_vehicles
                }, ensure_ascii=False)

            elif function_name == "get_vehicle_details":
                params = json.loads(function_args)
                vehicle_id = params.get("vehicle_id")
                
                if not vehicle_id:
                    return json.dumps({"error": "ID de vehículo requerido"})

                vehicle = await self._catalog_repository.get_vehicle_by_id(vehicle_id)
                
                if not vehicle:
                    return json.dumps({"error": "Vehículo no encontrado"})

                return json.dumps({
                    "id": vehicle.stock_id,
                    "marca": vehicle.brand,
                    "modelo": vehicle.model,
                    "año": vehicle.year,
                    "version": vehicle.version,
                    "precio": vehicle.price,
                    "kilometraje": vehicle.mileage,
                    "dimensiones": {
                        "largo": vehicle.length,
                        "ancho": vehicle.width,
                        "alto": vehicle.height
                    },
                    "caracteristicas": {
                        "bluetooth": vehicle.has_bluetooth,
                        "carplay": vehicle.has_carplay
                    }
                }, ensure_ascii=False)

            else:
                return json.dumps({"error": f"Función desconocida: {function_name}"})
        except Exception as ex:
            logger.error(f"Error in _execute_tool for {function_name}: {ex}")
            return json.dumps({"error": "Error interno al procesar la solicitud"})

    async def generate_structured_response(self, prompt: str, response_type: type):
        """
        Genera una respuesta estructurada del LLM según un esquema específico.
        
        Args:
            prompt: Prompt para el LLM
            response_type: Tipo de respuesta esperada
            
        Raises:
            NotImplementedError: Esta funcionalidad aún no está implementada
        """
        logger.warning("Structured response generation not yet implemented")
        raise NotImplementedError("Structured response generation not yet implemented")

File: baba-chatbot-net/src/Baba.Chatbot.Integrations/Llm/LlmClient.cs
using Baba.Chatbot.Application.Abstractions;
using Baba.Chatbot.Domain.ValueObjects;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using OpenAI.Chat;
using System.Text.Json;

namespace Baba.Chatbot.Integrations.Llm;

/// <summary>
/// Cliente para interactuar con modelos de lenguaje de OpenAI
/// Implementa generación de respuestas con RAG y function calling para búsqueda de vehículos
/// </summary>
public class LlmClient : ILlmClient
{
    private readonly ChatClient _chatClient;
    private readonly ILogger<LlmClient> _logger;
    private readonly KnowledgeRepository _knowledgeRepository;
    private readonly ICatalogRepository? _catalogRepository;
    private readonly string _model;
    private readonly float _temperature;

    /// <summary>
    /// Inicializa una nueva instancia del cliente LLM con configuración de OpenAI
    /// </summary>
    public LlmClient(IConfiguration configuration, ILogger<LlmClient> logger, KnowledgeRepository knowledgeRepository, ICatalogRepository? catalogRepository = null)
    {
        _logger = logger;
        _knowledgeRepository = knowledgeRepository;
        _catalogRepository = catalogRepository;

        var apiKey = configuration["OpenAI:ApiKey"] ?? throw new InvalidOperationException("OpenAI API Key is not configured");
        
        _model = configuration["OpenAI:Model"] ?? "gpt-4o-mini";
        _temperature = (float)configuration.GetValue<double>("OpenAI:Temperature", 0.7);

        _chatClient = new ChatClient(_model, apiKey);
        
        _logger.LogInformation("LlmClient initialized with model: {Model}", _model);
    }

    /// <summary>
    /// Genera una respuesta usando el LLM con contexto RAG y herramientas de búsqueda de vehículos
    /// Enriquece el prompt del sistema con información relevante de la base de conocimiento
    /// </summary>
    public async Task<string> GenerateResponseAsync(string systemPrompt, string userMessage,List<string>? conversationHistory = null,CancellationToken cancellationToken = default)
    {
        try
        {
            var relevantContext = _knowledgeRepository.SearchRelevantContext(userMessage, maxChunks: 3);
            
            var enhancedSystemPrompt = systemPrompt;
            if (!string.IsNullOrEmpty(relevantContext))
            {
                enhancedSystemPrompt = $"{systemPrompt}\n\n---\n\n## INFORMACIÓN DE CONTEXTO (Base de Conocimiento)\n\nUsa esta información para responder con datos precisos:\n\n{relevantContext}\n\n---\n\nIMPORTANTE: Usa SOLO la información del contexto cuando esté disponible. Si la información no está en el contexto, indica que necesitas verificar con un asesor.";
                _logger.LogInformation("Enhanced prompt with {Length} characters of context", relevantContext.Length);
            }

            var messages = new List<ChatMessage>
            {
                new SystemChatMessage(enhancedSystemPrompt)
            };

            if (conversationHistory != null && conversationHistory.Any())
            {
                for (int i = 0; i < conversationHistory.Count; i++)
                {
                    if (i % 2 == 0)
                    {
                        messages.Add(new UserChatMessage(conversationHistory[i]));
                    }
                    else
                    {
                        messages.Add(new AssistantChatMessage(conversationHistory[i]));
                    }
                }
            }

            messages.Add(new UserChatMessage(userMessage));

            _logger.LogDebug("Sending chat completion request with {MessageCount} messages", messages.Count);

            var options = new ChatCompletionOptions
            {
                Temperature = _temperature
            };

            if (_catalogRepository != null)
            {
                options.Tools.Add(ChatTool.CreateFunctionTool(
                    functionName: "search_vehicles",
                    functionDescription: "Busca vehículos en el inventario según criterios específicos como marca, modelo, precio, año, etc.",
                    functionParameters: BinaryData.FromString("""
                    {
                        "type": "object",
                        "properties": {
                            "brand": {
                                "type": "string",
                                "description": "Marca del vehículo (ej: Toyota, Honda, BMW)"
                            },
                            "model": {
                                "type": "string",
                                "description": "Modelo del vehículo (ej: Corolla, Civic, Serie 3)"
                            },
                            "min_price": {
                                "type": "number",
                                "description": "Precio mínimo en pesos"
                            },
                            "max_price": {
                                "type": "number",
                                "description": "Precio máximo en pesos"
                            },
                            "min_year": {
                                "type": "integer",
                                "description": "Año mínimo del vehículo"
                            },
                            "max_year": {
                                "type": "integer",
                                "description": "Año máximo del vehículo"
                            },
                            "max_mileage": {
                                "type": "integer",
                                "description": "Kilometraje máximo permitido"
                            }
                        },
                        "required": []
                    }
                    """
                    )
                ));

                options.Tools.Add(ChatTool.CreateFunctionTool(
                    functionName: "get_vehicle_details",
                    functionDescription: "Obtiene información detallada de un vehículo específico usando su ID o stock_id",
                    functionParameters: BinaryData.FromString("""
                    {
                        "type": "object",
                        "properties": {
                            "vehicle_id": {
                                "type": "string",
                                "description": "ID o stock_id del vehículo"
                            }
                        },
                        "required": ["vehicle_id"]
                    }
                    """
                    )
                ));
            }

            var completion = await _chatClient.CompleteChatAsync(messages, options, cancellationToken);
            
            if (completion.Value.FinishReason == ChatFinishReason.ToolCalls)
            {
                var response = await HandleToolCallsAsync(messages, completion.Value, options, cancellationToken);
                return response;
            }

            var responseText = completion.Value.Content[0].Text;
            
            _logger.LogInformation("Generated response: {ResponseLength} characters", responseText.Length);
            
            return responseText;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating LLM response");
            throw;
        }
    }

    /// <summary>
    /// Maneja las llamadas a herramientas (function calling) del LLM
    /// Ejecuta las funciones solicitadas y envía los resultados de vuelta al modelo
    /// </summary>
    private async Task<string> HandleToolCallsAsync(List<ChatMessage> messages,ChatCompletion completion,ChatCompletionOptions options,CancellationToken cancellationToken)
    {
        if (_catalogRepository == null)
        {
            return "Lo siento, no puedo consultar el catálogo en este momento.";
        }

        messages.Add(new AssistantChatMessage(completion));

        foreach (var toolCall in completion.ToolCalls)
        {
            var functionName = toolCall.FunctionName;
            var functionArgs = toolCall.FunctionArguments.ToString();

            _logger.LogInformation("Processing tool call: {FunctionName} with args: {Args}", functionName, functionArgs);

            string functionResult;
            try
            {
                functionResult = await ExecuteToolAsync(functionName, functionArgs, cancellationToken);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error executing tool {FunctionName}", functionName);
                functionResult = JsonSerializer.Serialize(new { error = "Error ejecutando la búsqueda" });
            }

            messages.Add(new ToolChatMessage(toolCall.Id, functionResult));
        }

        var finalCompletion = await _chatClient.CompleteChatAsync(messages, options, cancellationToken);
        
        return finalCompletion.Value.Content[0].Text;
    }

    /// <summary>
    /// Ejecuta una función/herramienta específica según su nombre
    /// Soporta búsqueda de vehículos y obtención de detalles
    /// </summary>
    private async Task<string> ExecuteToolAsync(string functionName, string functionArgs, CancellationToken cancellationToken)
    {
        if (_catalogRepository == null)
        {
            return JsonSerializer.Serialize(new { error = "Catálogo no disponible" });
        }

        try
        {
            switch (functionName)
            {
                case "search_vehicles":
                    var searchParams = JsonSerializer.Deserialize<SearchVehiclesParams>(functionArgs);
                    if (searchParams == null)
                    {
                        return JsonSerializer.Serialize(new { error = "Parámetros inválidos" });
                    }

                    var query = new VehicleQuery
                    {
                        Brand = searchParams.brand,
                        Model = searchParams.model,
                        MinPrice = searchParams.min_price,
                        MaxPrice = searchParams.max_price,
                        MinYear = searchParams.min_year,
                        MaxYear = searchParams.max_year,
                        MaxMileage = searchParams.max_mileage
                    };

                    var vehicles = await _catalogRepository.SearchVehiclesAsync(query, cancellationToken);
                    
                    var limitedVehicles = vehicles.Take(10).Select(v => new
                    {
                        id = v.StockId,
                        marca = v.Brand,
                        modelo = v.Model,
                        año = v.Year,
                        version = v.Version,
                        precio = v.Price,
                        kilometraje = v.Mileage,
                        dimensiones = new
                        {
                            largo = v.Length,
                            ancho = v.Width,
                            alto = v.Height
                        },
                        caracteristicas = new
                        {
                            bluetooth = v.HasBluetooth,
                            carplay = v.HasCarPlay
                        }
                    }).ToList();

                    return JsonSerializer.Serialize(new
                    {
                        total = vehicles.Count,
                        resultados_mostrados = limitedVehicles.Count,
                        vehiculos = limitedVehicles
                    });

                case "get_vehicle_details":
                    var detailParams = JsonSerializer.Deserialize<GetVehicleDetailsParams>(functionArgs);
                    if (detailParams?.vehicle_id == null)
                    {
                        return JsonSerializer.Serialize(new { error = "ID de vehículo requerido" });
                    }

                    var vehicle = await _catalogRepository.GetVehicleByIdAsync(detailParams.vehicle_id, cancellationToken);
                    
                    if (vehicle == null)
                    {
                        return JsonSerializer.Serialize(new { error = "Vehículo no encontrado" });
                    }

                    return JsonSerializer.Serialize(new
                    {
                        id = vehicle.StockId,
                        marca = vehicle.Brand,
                        modelo = vehicle.Model,
                        año = vehicle.Year,
                        version = vehicle.Version,
                        precio = vehicle.Price,
                        kilometraje = vehicle.Mileage,
                        dimensiones = new
                        {
                            largo = vehicle.Length,
                            ancho = vehicle.Width,
                            alto = vehicle.Height
                        },
                        caracteristicas = new
                        {
                            bluetooth = vehicle.HasBluetooth,
                            carplay = vehicle.HasCarPlay
                        }
                    });

                default:
                    return JsonSerializer.Serialize(new { error = $"Función desconocida: {functionName}" });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error en ExecuteToolAsync para {FunctionName}", functionName);
            return JsonSerializer.Serialize(new { error = "Error interno al procesar la solicitud" });
        }
    }

    /// <summary>
    /// Genera una respuesta estructurada del LLM según un esquema específico
    /// </summary>
    public Task<T> GenerateStructuredResponseAsync<T>(string prompt, CancellationToken cancellationToken = default) where T : class
    {
        _logger.LogWarning("Structured response generation not yet implemented");
        throw new NotImplementedException("Structured response generation not yet implemented");
    }

    /// <summary>
    /// DTO para deserializar parámetros de búsqueda de vehículos
    /// </summary>
    private class SearchVehiclesParams
    {
        public string? brand { get; set; }
        public string? model { get; set; }
        public decimal? min_price { get; set; }
        public decimal? max_price { get; set; }
        public int? min_year { get; set; }
        public int? max_year { get; set; }
        public int? max_mileage { get; set; }
    }

    /// <summary>
    /// DTO para deserializar parámetros de obtención de detalles de vehículo
    /// </summary>
    private class GetVehicleDetailsParams
    {
        public string? vehicle_id { get; set; }
    }
}

